File: windows/BlueLink.App/Usb/MtpFileCipher.cs
using System.Buffers.Binary;
using System.Security.Cryptography;

namespace BlueLink.Usb;

/// <summary>BLM1: bounded, independently authenticated records. Keys are fresh per attempt and sent only over BTX.</summary>
public static class MtpFileCipher
{
    public const int ChunkSize = 1024 * 1024;
    public static long EncodedSize(long size)
    {
        ValidateSize(size);
        return checked(size + 16 * Math.Max(1, (size + ChunkSize - 1) / ChunkSize));
    }

    public static async Task EncryptAsync(Stream input, Stream output, Guid id, long size, byte[] key,
        Func<Task>? checkpoint = null, CancellationToken token = default)
    {
        Validate(size, key);
        using var cipher = new ChaCha20Poly1305(key);
        var plain = new byte[ChunkSize];
        var encrypted = new byte[ChunkSize];
        var tag = new byte[16];
        long offset = 0;
        try
        {
        for (long index = 0; index < Math.Max(1, (size + ChunkSize - 1) / ChunkSize); index++)
        {
            token.ThrowIfCancellationRequested();
            if (checkpoint is not null) await checkpoint();
            var length = (int)Math.Min(ChunkSize, size - offset);
            await input.ReadExactlyAsync(plain.AsMemory(0, length), token);
            cipher.Encrypt(Nonce(index), plain.AsSpan(0, length), encrypted.AsSpan(0, length), tag,
                Aad(id, size, index, length));
            await output.WriteAsync(encrypted.AsMemory(0, length), token);
            await output.WriteAsync(tag, token);
            offset += length;
        }
        if (input.ReadByte() != -1) throw new InvalidDataException("USB source size changed");
        }
        finally { CryptographicOperations.ZeroMemory(plain); }
    }

    public static async Task DecryptAsync(Stream input, Guid id, long size, byte[] key,
        Func<ReadOnlyMemory<byte>, Task> accept, Func<Task>? checkpoint = null, CancellationToken token = default)
    {
        Validate(size, key);
        using var cipher = new ChaCha20Poly1305(key);
        var plain = new byte[ChunkSize];
        var encrypted = new byte[ChunkSize];
        var tag = new byte[16];
        long offset = 0;
        try
        {
            for (long index = 0; index < Math.Max(1, (size + ChunkSize - 1) / ChunkSize); index++)
            {
                token.ThrowIfCancellationRequested();
                if (checkpoint is not null) await checkpoint();
                var length = (int)Math.Min(ChunkSize, size - offset);
                await input.ReadExactlyAsync(encrypted.AsMemory(0, length), token);
                await input.ReadExactlyAsync(tag, token);
                cipher.Decrypt(Nonce(index), encrypted.AsSpan(0, length), tag, plain.AsSpan(0, length),
                    Aad(id, size, index, length));
                await accept(plain.AsMemory(0, length));
                offset += length;
            }
            if (input.ReadByte() != -1) throw new InvalidDataException("Unexpected USB file suffix");
        }
        finally { CryptographicOperations.ZeroMemory(plain); }
    }

    private static void Validate(long size, byte[] key)
    {
        ValidateSize(size);
        if (key.Length != 32)
            throw new InvalidDataException("Invalid USB file parameters");
    }
    private static void ValidateSize(long size)
    {
        if (size < 0 || size > (long)ChunkSize * uint.MaxValue)
            throw new InvalidDataException("Invalid USB file size");
    }
    private static byte[] Nonce(long index)
    {
        var bytes = new byte[12];
        BinaryPrimitives.WriteInt32BigEndian(bytes, 0x424C4D31);
        BinaryPrimitives.WriteInt64BigEndian(bytes.AsSpan(4), index);
        return bytes;
    }
    private static byte[] Aad(Guid id, long size, long index, int length)
    {
        var bytes = new byte[40];
        BinaryPrimitives.WriteInt32BigEndian(bytes, 0x424C4D31);
        Convert.FromHexString(id.ToString("N")).CopyTo(bytes, 4);
        BinaryPrimitives.WriteInt64BigEndian(bytes.AsSpan(20), size);
        BinaryPrimitives.WriteInt64BigEndian(bytes.AsSpan(28), index);
        BinaryPrimitives.WriteInt32BigEndian(bytes.AsSpan(36), length);
        return bytes;
    }
}

File: protocol-core/src/main/java/com/bluelink/core/MtpFileCipher.java
package com.bluelink.core;

import javax.crypto.Cipher;
import javax.crypto.spec.IvParameterSpec;
import javax.crypto.spec.SecretKeySpec;
import java.io.*;
import java.nio.ByteBuffer;
import java.security.GeneralSecurityException;
import java.util.Arrays;
import java.util.UUID;

/** BLM1 interoperable authenticated spool, never stores a key in the public directory. */
public final class MtpFileCipher {
    public static final int CHUNK_SIZE = 1024 * 1024;
    @FunctionalInterface public interface Checkpoint { void run() throws IOException; }
    @FunctionalInterface public interface Sink { void accept(byte[] bytes, int length) throws IOException; }
    public static long encodedSize(long size) {
        if (size < 0 || size > (long)CHUNK_SIZE * 0xffffffffL)
            throw new IllegalArgumentException("Invalid USB file size");
        return Math.addExact(size, 16 * Math.max(1, (size + CHUNK_SIZE - 1) / CHUNK_SIZE));
    }
    public static void encrypt(InputStream input, OutputStream output, UUID id, long size, byte[] key,
                               Checkpoint checkpoint) throws IOException, GeneralSecurityException {
        validate(size, key);
        long offset = 0;
        byte[] buffer = new byte[CHUNK_SIZE];
        try {
            for (long index = 0; index < Math.max(1, (size + CHUNK_SIZE - 1) / CHUNK_SIZE); index++) {
                checkpoint.run();
                int length = (int)Math.min(CHUNK_SIZE, size - offset);
                new DataInputStream(input).readFully(buffer, 0, length);
                Cipher cipher = cipher(Cipher.ENCRYPT_MODE, key, id, size, index, length);
                output.write(cipher.doFinal(buffer, 0, length));
                offset += length;
            }
            if (input.read() != -1) throw new IOException("USB source size changed");
        } finally { Arrays.fill(buffer, (byte)0); }
    }
    public static void decrypt(InputStream input, UUID id, long size, byte[] key, Sink sink,
                               Checkpoint checkpoint) throws IOException, GeneralSecurityException {
        validate(size, key);
        long offset = 0;
        DataInputStream data = new DataInputStream(input);
        for (long index = 0; index < Math.max(1, (size + CHUNK_SIZE - 1) / CHUNK_SIZE); index++) {
            checkpoint.run();
            int length = (int)Math.min(CHUNK_SIZE, size - offset);
            byte[] encrypted = new byte[length + 16];
            data.readFully(encrypted);
            byte[] plain = cipher(Cipher.DECRYPT_MODE, key, id, size, index, length).doFinal(encrypted);
            try { sink.accept(plain, length); } finally { Arrays.fill(plain, (byte)0); }
            offset += length;
        }
        if (data.read() != -1) throw new IOException("Unexpected USB file suffix");
    }
    private static Cipher cipher(int mode, byte[] key, UUID id, long size, long index, int length)
            throws GeneralSecurityException {
        byte[] nonce = ByteBuffer.allocate(12).putInt(0x424c4d31).putLong(index).array();
        byte[] aad = ByteBuffer.allocate(40).putInt(0x424c4d31)
                .putLong(id.getMostSignificantBits()).putLong(id.getLeastSignificantBits())
                .putLong(size).putLong(index).putInt(length).array();
        Cipher cipher = CryptoProviders.chacha20Poly1305();
        cipher.init(mode, new SecretKeySpec(key, "ChaCha20"), new IvParameterSpec(nonce));
        cipher.updateAAD(aad);
        return cipher;
    }
    private static void validate(long size, byte[] key) throws IOException {
        if (size < 0 || size > (long)CHUNK_SIZE * 0xffffffffL || key.length != 32)
            throw new IOException("Invalid USB file parameters");
    }
    private MtpFileCipher() {}
}
